### ISSUE TITLE
spack external add

### ISSUE THREAD

**Issue by mdorier** [LLNL/spack#29811]
### Summary

The `spack external add` subcommand would be used to add an external package to packages.yaml or to an environment's yaml file (or any other scope via its `--scope` parameter) via the command line instead of editing the yaml file manually. 

Contrary to `spack external find`, this command would not be restricted to packages that are "detectable" (ie. not just those in listed `spack external list`). This command would take additional parameters, in particular `--prefix` or `--modules`, to specify where the package is located.

Note: an alternative would be to extend `spack external find` to allow supplying the prefix or modules, but I think using `add` makes it clear that spack is not trying to check that the package is indeed present.

### Rationale

At present, the vast majority of packages don't have a way of being detected via `spack external find`. Yet most supercomputers provide packages in the form of modules, and the only way to use these modules in a spack environment is to manually edit yaml files.

In the context of scripted builds, calling
```
spack external add something@1.2.3%comp@4.5.6 --modules something/1.2.3,comp/4.5.6
```
or
```
spack external add something@1.2.3%comp@4.5.6 --prefix /path/to/something
```
is much easier than editing a yaml file from e.g. a bash script.

### Additional information

Version of spack: 0.17.1-1070-5208fcf1ea

### General information

- [X] I have run `spack --version` and reported the version of Spack
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by adamjstewart** [comment]
`spack compiler add` and `spack compiler find` are aliases of each other, I don't think it would be a good idea to give `spack external add` and `spack external find` different behavior.

**Comment by mdorier** [comment]
Good point, but it could be another name then.

**Comment by alalazo** [comment]
Personal opinion, but I'd not be keen on giving users handles to defy more easily Spack software model (externals right now are essentially that). The rationale is that I think we should go in the direction of having externals which are detected as concrete specs with dependencies and automatically registered in the DB, since that would allow us to treat externals and other sources of already built/installed software[^1] uniformly.

As for scripting: how often do module changes on the HPC machines you target? Asking to try to understand what would be the benefit of implementing what is suggested above that rather than, say, start with a template `spack.yaml` tuned for the environment in which the script will run.

Finally, I wonder if there's anything missing in the support for [custom Spack extensions](https://spack.readthedocs.io/en/latest/extensions.html) to allow users to code what is described and extend Spack. Somehow, I don't think this should go in core Spack, but I think Spack should allow users to code this functionality by themselves easily enough.

[^1]: like Spack stores or buildcaches

**Comment by mdorier** [comment]
I'm working on some machines on which the admins do add and remove modules somewhat frequently. If I don't touch a project for a couple of months on these machines, I can expect my spack.yaml file to be outdated, and my installation script to break the next time I try building things. Hence my attempt to script the detection of modules.

**Comment by alalazo** [comment]
How do you get the module names? In your example you have:
```console
spack external add something@1.2.3%comp@4.5.6 --modules something/1.2.3,comp/4.5.6
```
and I assume in the script the literals `something/1.2.3` and `comp/4.5.6` are computed somehow (otherwise you have to edit the script which is the same effort as editing the `spack.yaml`). What I was wondering is if a template `spack.yaml` can serve the same purpose (think of Jinja2 as an example that comes bundled with Spack, but the mechanism may be anything suitable).

**Comment by mdorier** [comment]
Good point. My thought process was that `spack external find` either fails because the package has no detection procedure, or gives a prefix-based answer once the module is loaded, which is not ideal when the module requires other modules or does something else in addition to setting PATH. I thought having a functionalities that allowed me to provide `--modules` would solve this: no need for detection, and now the `modules` field of the package is used instead of the `prefix` field. I haven't thought much yet about how I would detect a module change, how I would figure out the full spec, the module name, etc.

Ideally I was envisioning using a simple text file listing lines like `<spec> => <module1>,<module2>,...`, from which I could call `spack external add spec --modules module1,module2...`. The alternative being, indeed, to use jinja and a template.
